src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,392 @@
+    scores.each_slice() += arma::repmat(keyPaddingMask, tgtSeqLen, 1);
+  }
+
+  for (size_t i = 0; i < numHeads * batchSize; ++i)
+  {
+    softmax.Forward(scores.slice(i), softmax.OutputParameter());
+    scores.slice(i) = softmax.OutputParameter();
+  }
+
+  //! Calculate the attention output i.e. matrix multiplication of softmax
+  //! output and vProj.
+  //! The shape of attnOutput : (tgtSeqLen, headDim, numHeads * batchSize).
+  attnOut = math::MultiplyCube2Cube(scores, vProj, false, false);
+
+  //! Now we will concatenate output of all the heads i.e. we will reshape
+  //! attnOut to (tgtSeqLen, embedDim, batchSize).
+  attnOut.reshape(tgtSeqLen, embedDim, batchSize);
+
+  //! The final output is the linear projection of attention output.
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    output.col(i) = arma::vectorise(attnOut.slice(i) * outWt
+        + arma::repmat(outBias, 1, tgtSeqLen).t());
+  }
+}
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+template <typename eT>
+void MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+Backward(const arma::Mat<eT>& /* input */,
+         const arma::Mat<eT>& gy,
+         arma::Mat<eT>& g)
+{
+  typedef typename arma::Cube<eT> CubeType;
+
+  const size_t tgtSeqLen = gy.n_rows / embedDim;
+  const size_t batchSize = gy.n_cols;
+
+  //! Reshape the propagated gradient into a cube.
+  //! The shape of gyTemp : (tgtSeqLen, embedDim, batchSize).
+  //! We need not split it into n heads now because this is the part when
+  //! output were concatenated from n heads.
+  CubeType gyTemp(const_cast<arma::Mat<eT>&>(gy).memptr(), tgtSeqLen, embedDim,
+      batchSize, true, false);
+
+  //! The shape of gyTemp : (tgtSeqLen, embedDim, batchSize).
+  //! The shape of outWt : (embedDim, embedDim).
+  gyTemp = math::MultiplyCube2Mat(gyTemp, outWt, false, true);
+
+  //! Now since the shape of gyTemp is (tgtSeqLen, embedDim, batchSize). We will
+  //! split it into n heads.
+  //! The shape of gyTemp : (tgtSeqLen, headDim, numHeads * batchSize).
+  gyTemp.reshape(tgtSeqLen, headDim, numHeads * batchSize);
+
+  //! The shape of gyTemp : (tgtSeqLen, headDim, numHeads * batchSize).
+  //! The shape of vProj : (srcSeqLen, headDim, numHeads * batchSize).
+  //! So the new shape of gyTemp : (tgtSeqLen, srcSeqLen, numHeads * batchSize).
+  gyTemp = math::MultiplyCube2Cube(gyTemp, vProj, false, true);
+
+  for (size_t i = 0; i < numHeads * batchSize; ++i)
+  {
+    //! We will perform backpropagation of softmax over each slice of gyTemp.
+    softmax.Backward(scores.slice(i), gyTemp.slice(i), gyTemp.slice(i));
+  }
+
+  //! The shape of kProj : (srcSeqLen, headDim, numHeads * batchSize).
+  //! The shape of gyTemp : (tgtSeqLen, srcSeqLen, numHeads * batchSize).
+  //! The new shape of gyTemp : (tgtSeqLen, headDim, numHeads * batchSize).
+  gyTemp = math::MultiplyCube2Cube(gyTemp, kProj) / std::sqrt(headDim);
```

Comment:
I saw Jacobian Test fails for `query = key = value` (i.e. Self-Attention case) but the reason is "how backward pass is tested?".

Considering the case when the key or value is not provided. Then key and value will be extracted from the input (i.e. query), but the thing is key and value are not input themselves. They are just "equal" to the input. When the Jacobian Test is held, f(x + h) and f(x - h) are calculated. It is expected that (x + h) or (x - h) only changes the input but since we are extracting key and value from the input itself, key and value also get mutated.

Consider this example--

![first_principle](https://user-images.githubusercontent.com/35535378/89489838-fdd13a00-d7c8-11ea-9bba-eece81e09e2f.gif)

Maybe this will be solved if we restrict the user to provide key and value even if they are equal to input (i.e. query) like other libraries do. What do you say??

Reply:
Are you telling me that `key` and `value` are constants (I mean they won't change during training)? I think that's not true in case of e.g. Self-Attention.

So, `key` and `value` are variables. If these variables change during training, you can't treat them as constants. In this case you have to calculate their partial derivatives too.

I totally agree with your example. But in this case `alpha` doesn't depend on `x`, it's a different variable.

Reply:
The thing is backward pass doesn't flow through `key` or `query` i.e. dL/d(key) or dL/d(value) is not calculated and if it is calculated, what will we do with them?? For the self-attention case, after every forward pass `key` and `value` are set equal to `query`.

>But in this case alpha doesn't depend on x, it's a different variable.

In my opinion, the same is the case here also. Again the same example--

![example](https://user-images.githubusercontent.com/35535378/89511460-91b5fc80-d7ef-11ea-960d-5a1aca66a65f.gif)

I locally tested with `query = key = value` and it passed. Let me push the changes to explain.

Reply:
> The thing is backward pass doesn't flow through key or query

I don't understand why. How are you going to obtain `key` and `query`? I thought you calculate them by forward propagation. So, they're not independent external parameters. They're variables like `query`. In this case you have to calculate `dL/d(query)`, `dL/d(key)`, and `dL/d(value)` and then calculate the gradient using these partial derivatives. But, in fact you use only `dL/d(query)`.

> I locally tested with query = key = value and it passed.

I still think these formulas aren't correct in strict mathematical sense. But, the difference could be small, and we could neglect it. Perhaps, there's something I don't understand correctly. Could you send me the link to any other implementation?

Reply:
Oops...The correct statement is "The thing is backward pass doesn't flow through key or <strike>query</strike> value". I corrected the above comment also.

Suppose there is `gy`. We get `dL/d(query)`, `dL/d(key)`, and `dL/d(value)` from it. So what should be `g` in terms of these partial derivatives??
What I understand is-- For the Backpropagation, we should just focus on `dL/d(input)` and in this case, it is `dL/d(query)`.

Unfortunately, I don't have implementation for Backpropagation of the multihead attention layer of any other library. For the forward propagation--
i) https://www.tensorflow.org/tutorials/text/transformer#multi-head_attention
ii) https://github.com/pytorch/pytorch/blob/c30bc6d4d7b54296179d717c15cd7507ba520bae/torch/nn/functional.py#L3865

Reply:
Yeah, I meant `value` too.

> What I understand is-- For the Backpropagation, we should just focus on dL/d(input)

Yes, you're right. But why aren't `key` and `value` parts of the input?

> Unfortunately, I don't have implementation for Backpropagation of the multihead attention layer of any other library.

The libraries do it automatically, you just need to build a model using the required blocks. I think we could compare the gradients of our implementation and of the tensorflow implementation. I think this'll solve the question.

Reply:
>I think we could compare the gradients of our implementation and of the tensorflow implementation.

So looks like everything boils down to comparing the Forward function because I (and all of us :D) trust the `CheckGradient` and `JacobianTest` functions in `ann_test_tools.cpp`. Also, I can see that the tolerance (1e-06) in the Gradient check of Multihead Attention is the lowest of all gradient checks in `ann_layer_test.cpp`.

About comparing Forward function, I will respond before going to sleep today :)

Reply:
> > What I understand is-- For the Backpropagation, we should just focus on dL/d(input)

> Yes, you're right. But why aren't key and value parts of the input?

Well, technically they're not parts of the input since you pass them in a different way. But in fact they have the same meaning as `query`, the MultiHeadAttention layer obtains them in the same way by means of the forward pass. Therefore I think they should be handled in the same way as `query`.

Reply:
> About comparing Forward function, I will respond before going to sleep today :)

I think the `Forward()` function is correct.

> Also, I can see that the tolerance (1e-06) in the Gradient check of Multihead Attention is the lowest of all gradient checks in ann_layer_test.cpp.

As I said perhaps, the difference is too small. So, maybe we could neglect it.

Reply:
>Therefore I think they should be handled in the same way as query.

So what do you suggest about this? 
I already tried different ways for this if you remember. Since I rebased and squashed all previous commits, I can't show them now. The various methods I already tried apart from current implementation--
i) Concatenate `query`, `key` and `value` as a single input. Disadvantage-- deliberate copy even if not required, Jacobian Test fails always.
ii) Policy-based input. Disadvantage-- had to implement multiple versions of class methods depending on Policy.
iii) The `Concat` layer (not `Concatenate`). I realized later, it's not how things are done in multihead attention.

Reply:
> So what do you suggest about this?

I think I have an idea. Let me a couple of days to write a prototype. Do you have other work to do? Can you leave this PR for a couple of days? I mean is there any other layers that aren't blocked by `MultiHeadAttention`?

Reply:
>Let me a couple of days to write a prototype.

That's great :)

>Do you have other work to do?

Yup, I have to work on Positional Encoding and Label Smoothing. Linear3D (PR 2508) is also ready. I think that's enough for two days :)

Reply:
>>What I understand is-- For the Backpropagation, we should just focus on dL/d(input)

>Yes, you're right. But why aren't key and value parts of the input?

@lozhnikov: This [answer from cross-validated](https://stats.stackexchange.com/a/424127) can help us understand this. It's true that not much explanation is there in the paper about what actually are `query`, `key`, and `value`?